commands/build/provision.go

```diff
@@ -75,133 +89,603 @@ func (recv *ProvisionStackCmd) Execute(args []string) bool {
 
-		for _, regionOutput := range stackOutput.Regions {
-			go provisionStackPoll(environment, regionOutput, outputChan, failureChan)
+					hotswapChan <- hotswapData
+				} else {
+
+					failureChan <- struct{}{}
+				}
+
+			}(environment, region)
 		}
 
-		for i := 0; i < regionCount; i++ {
+		// provision if any region's stack is too old
+		shouldHotswap := true
+
+		for i := 0; i < len(environment.Regions); i++ {
 			select {
-			case regionOutput := <-outputChan:
-				provisionOutput.Regions = append(provisionOutput.Regions, regionOutput)
+			case hotswapData := <-hotswapChan:
+
+				switch hotswapData.stackStatus {
+				case "CREATE_COMPLETE", "UPDATE_COMPLETE":
+					// only states eligible for hot swap
+
+				case "CREATE_FAILED",
+					"DELETE_COMPLETE",
+					"DELETE_FAILED",
+					"DELETE_IN_PROGRESS",
+					"ROLLBACK_COMPLETE",
+					"ROLLBACK_FAILED",
+					"ROLLBACK_IN_PROGRESS":
+
+					hotswapData.shouldHotswap = false
+
+				case "CREATE_IN_PROGRESS":
+
+					log.Error("Can't hot swap a stack being created")
+					return
+
+				case "UPDATE_COMPLETE_CLEANUP_IN_PROGRESS",
+					"UPDATE_IN_PROGRESS":
+
+					log.Error("A previous hot swap is still in progress",
+						"StackStatus", hotswapData.stackStatus)
+					return
+
+				case "UPDATE_ROLLBACK_COMPLETE",
+					"UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS",
+					"UPDATE_ROLLBACK_FAILED",
+					"UPDATE_ROLLBACK_IN_PROGRESS":
+
+					log.Error("A previous hot swap appears to have failed",
+						"StackStatus", hotswapData.stackStatus)
+					return
+
+				default:
+
+					log.Error("Undocumented stack status",
+						"StackStatus", hotswapData.stackStatus)
+					return
+				}
+
+				shouldHotswap = shouldHotswap && hotswapData.shouldHotswap
+				hotswapStructs = append(hotswapStructs, hotswapData)
+
 			case _ = <-failureChan:
-				commandSuccess = false
+				return
 			}
 		}
 
-		if commandSuccess {
+		if shouldHotswap {
+			success = HotswapStack(log, config, environment, hotswapStructs)
+		} else {
+			success = ProvisionStack(log, config, environment)
+		}
 
-			provisionBytes, err := json.Marshal(provisionOutput)
-			if err != nil {
-				log.Error("json.Marshal", "Error", err)
-				os.Exit(1)
+	} else {
+
+		success = ProvisionStack(log, config, environment)
+	}
+
+	return
+}
+
+func checkShouldHotswapRegion(log log15.Logger, environment *conf.Environment,
+	region *conf.Region) (hotswapData hotswapStruct, success bool) {
+
+	log = log.New("Region", region.Name)
+
+	log.Debug("checkShouldHotswapRegion BEGIN")
+	defer log.Debug("checkShouldHotswapRegion END")
+
+	hotswapData.shouldHotswap = true
+	hotswapData.region = region.Name
+
+	roleARN, err := environment.GetRoleARN(region.Name)
+	if err != nil {
+		log.Error("GetRoleARN", "Error", err)
+		return
+	}
+
+	roleSession := aws_session.STS(region.Name, roleARN, 0)
+
+	elbClient := elb.New(roleSession)
+	cfnClient := cloudformation.New(roleSession)
+
+elbLoop:
```

Comment:
I don't think I'll ever get used to seeing labels/goto in Go.  I'm having flashbacks to VB.